go/pipeline/batch.go

```diff
@@ -0,0 +1,396 @@
+//
+// Parameters:
+//
+//	timeout - The timeout in milliseconds.
+//
+// Returns:
+//
+//	The updated ClusterBatchOptions instance.
+func (cbo *ClusterBatchOptions) WithTimeout(timeout uint32) *ClusterBatchOptions {
+	cbo.Timeout = &timeout
+	return cbo
+}
+
+// TODO ensure only single node route is allowed (use config.NotMultiNode?)
+
+// WithRoute sets the routing strategy for the batch.
+//
+// Parameters:
+//
+//	route - The routing strategy to use.
+//
+// Returns:
+//
+//	The updated ClusterBatchOptions instance.
+func (cbo *ClusterBatchOptions) WithRoute(route config.Route) *ClusterBatchOptions {
+	cbo.Route = &route
+	return cbo
+}
+
+// WithRetryStrategy sets the retry strategy for the batch.
+//
+// Parameters:
+//
+//	retryStrategy - The retry strategy to use.
+//
+// Returns:
+//
+//	The updated ClusterBatchOptions instance.
+func (cbo *ClusterBatchOptions) WithRetryStrategy(retryStrategy ClusterBatchRetryStrategy) *ClusterBatchOptions {
+	cbo.RetryStrategy = &retryStrategy
+	return cbo
+}
+
+// ====================
+
+// TODO - move this struct and convert methods to internals
+type BatchOptions struct {
+	Timeout       *uint32
+	Route         *config.Route
+	RetryStrategy *ClusterBatchRetryStrategy
+}
+
+func (sbo StandaloneBatchOptions) Convert() BatchOptions {
+	return BatchOptions{Timeout: sbo.Timeout}
+}
+
+func (cbo ClusterBatchOptions) Convert() BatchOptions {
+	return BatchOptions{Timeout: cbo.Timeout, Route: cbo.Route, RetryStrategy: cbo.RetryStrategy}
+}
+
+// ====================
+
+// TODO make private if possible
+type Batch struct {
+	Commands []Cmd
+	IsAtomic bool
+}
+
+// BaseBatch is the base structure for both standalone and cluster batch implementations.
+type BaseBatch[T StandaloneBatch | ClusterBatch] struct {
+	Batch
+	self *T
+}
+
+// StandaloneBatch is the batch implementation for standalone Valkey servers.
+// Batches allow the execution of a group of commands in a single step.
+//
+// Batch Response:
+//
+//	An array of command responses is returned by the client exec command, in the order they were given.
+//	Each element in the array represents a command given to the batch.
+//	The response for each command depends on the executed Valkey command.
+//	Specific response types are documented alongside each method.
+type StandaloneBatch struct {
+	BaseBatch[StandaloneBatch]
+}
+
+// ClusterBatch is the batch implementation for clustered Valkey servers.
+// Batches allow the execution of a group of commands in a single step.
+//
+// Batch Response:
+//
+//	An array of command responses is returned by the client exec command, in the order they were given.
+//	Each element in the array represents a command given to the batch.
+//	The response for each command depends on the executed Valkey command.
+//	Specific response types are documented alongside each method.
+type ClusterBatch struct {
+	BaseBatch[ClusterBatch]
+}
+
+// ====================
+
+func (b Batch) Convert(response []any) ([]any, error) {
```

Comment:
Documentation?

Reply:
We need to move struct `Batch` and this method to internals, so they won't be a part of public API.
Will be done in a follow-up PR.